### PR TITLE
[ENH] Draw nicer arrows, shorten the edges

### LINKED ISSUE #174: Directed Network: have a nicer self-directed visualization
##### Network version

1.5.1

##### Orange version

3.28.dev

##### Expected behavior
Self-directed nodes have a clear return loop.
K9Dz3

##### Actual behavior
What that?
Screen Shot 2020-10-23 at 14 47 48

##### Steps to reproduce the behavior
[CIRC.NET.zip](https://github.com/biolab/orange3-network/files/5429655/CIRC.NET.zip)

### PR DESCRIPTION
**Pull request by janezd**
Fixes #174: draw loops.

Screenshot 2021-01-28 at 18 54 07

Besides, edges, in particular with arrows, are ugly. This PR makes them shorter, nicer.

Screenshot 2020-10-23 at 15 29 05

Screenshot 2021-01-22 at 19 02 21

No tests, it's all in `paint`.
##### Includes
- [X] Code changes

### CODE REVIEW

File: orangecontrib/network/widgets/OWNxExplorer.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ class Inputs:
 
     class Outputs(OWDataProjectionWidget.Outputs):
         subgraph = Output("Selected sub-network", Network)
-        unselected_subgraph = Output("Remaining sub-network", Network)
+        #unselected_subgraph = Output("Remaining sub-network", Network)
```
Comment (ajdapretnar) [inline orangecontrib/network/widgets/OWNxExplorer.py:35]: Is there a reason why this is commended out?

Comment (janezd) [reply orangecontrib/network/widgets/OWNxExplorer.py:35]: Yes. Together with the inherited outputs, this widget would have five, which doesn't work until merging https://github.com/biolab/orange-canvas-core/pull/168 and releasing new canvas. :)

Comment (janezd) [reply orangecontrib/network/widgets/OWNxExplorer.py:35]: Actually, I'm gonna merge https://github.com/biolab/orange-canvas-core/pull/168 now, and then uncomment this.

Comment (janezd) [reply orangecontrib/network/widgets/OWNxExplorer.py]: I removed the commit that commented this out.

**Comment by ajdapretnar** [comment]
Eh, I accidentally commented in the issue instead of PR. 🤦‍♀️ Moving the comments here.

1. Right. This is very subtle. I like it, but I doubt anyone will notice it? 🤔 It doesn't bother though, so we can leave it.
2. a. Yes, edges should touch. Always, IMO.
2. b. It was nicer before. I would keep it as is.

**Comment by janezd** [comment]
1. Nobody will notice. I'd keep it just because it's difficult to make them shorter.
2. a. You don't like the last of the above images?

Touching is difficult.

Round caps, touching: kind of works. It's touching, though not really.

Screenshot 2021-01-29 at 12 27 29

Flat caps, touching: can't look nice for wider edges.

Screenshot 2021-01-29 at 12 35 05

Making the lines a bit longer doesn't work, either - they noticeably overlap with circles.

If you agree, I'd go either with touching round caps. Always, if you want.

3. What do you mean with "before"? Before what?

**Comment by ajdapretnar** [comment]
2. Round caps are ok. They indeed look the best.
3. I meant on master. My issue is the following: scaling / zooming is off. On master, edges are scaled accordingly. On this branch this is what happens (in sequential order):
Screen Shot 2021-01-29 at 13 53 58
Screen Shot 2021-01-29 at 13 54 04
Screen Shot 2021-01-29 at 13 54 10
Screen Shot 2021-01-29 at 13 54 16
Edges are just not right. Sometimes they are too small, sometimes to big and overlap. I tried this and it doesn't happen on master.

**Comment by janezd** [comment]
Could you please change

```python
        sizes0, sizes1 = self.sizes[::2] + 3, self.sizes[1::2] + 3
        sizes0 += w3
        sizes1 += w3
```

to

```python
        sizes0, sizes1 = self.sizes[::2] + w3, self.sizes[1::2] + w3
```

(note `3` -> `w3` in the remaining line!) and check whether this now works OK?

Edges will touch and should not scale incorrectly. They will still overlap, but only for ovelapping vertices - where this is the best and only reasonable thing to do.

If you confirm this is OK, I'll change the commit. Thanks!

**Comment by ajdapretnar** [comment]
Yes, this is better.

**Comment by thocevar** [comment]
This crashes on `mips_c2_cp_leu`.

```
orange3-network\orangecontrib\network\widgets\graphview.py", line 145, in paint
    rxs = xs - ds * fx
ValueError: operands could not be broadcast together with shapes (13,) (134,)
```

Do we have any directed networks in the example network dataset?

**Comment by ajdapretnar** [comment]
I use circ.net.

[CIRC.NET.zip](https://github.com/biolab/orange3-network/files/5909782/CIRC.NET.zip)

**Comment by janezd** [comment]
> This crashes on `mips_c2_cp_leu`.

Fixed.

**Comment by ajdapretnar** [comment]
Oh boy, I like this fix, but I found something else. This is CIRC.NET. It seems like auto-directed node edge is resized on node size not edge width. And it seems to have an artifact beside it. I am sorry to complicate things over and over.

Screen Shot 2021-02-19 at 15 35 33

**Comment by janezd** [comment]
The fat edge works as expected! After much debuging (within which I hunted for a bug in pajek reader and then in network implementation) I discovered the that loop appears twice in the file and thus has a combined weight of 2.

The artefact was a result of an actual that I now fixed.

**Pull request merged**